include/bluetooth/conn_ctx.h

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (c) 2018 Nordic Semiconductor ASA
+ *
+ * SPDX-License-Identifier: LicenseRef-BSD-5-Clause-Nordic
+ */
+
+/**
+ * @file
+ * @defgroup bt_conn_ctx Bluetooth connection context library API
+ * @{
+ * @brief API for the Bluetooth connection context library.
+ */
+
+#ifndef BT_CONN_CTX_H_
+#define BT_CONN_CTX_H_
+
+#include <zephyr.h>
+#include <misc/__assert.h>
+#include <bluetooth/conn.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**@brief Macro for defining a Bluetooth connection context library instance.
+ *
+ * @param  _name	Name of the instance.
+ * @param  _max_clients	Maximum number of clients connected at a time.
+ * @param  _ctx_sz	Context size in bytes for a single connection.
+ */
+#define BT_CONN_CTX_DEF(_name, _max_clients, _ctx_sz)                          \
+	K_MEM_SLAB_DEFINE(_name##_mem_slab,                                    \
+			  ROUND_UP(_ctx_sz, CONFIG_BT_CONN_CTX_MEM_BUF_ALIGN), \
+			  (_max_clients),                                      \
+			  CONFIG_BT_CONN_CTX_MEM_BUF_ALIGN);                   \
+	K_MUTEX_DEFINE(_name##_mutex);                                         \
+	static struct bt_conn_ctx_lib CONCAT(_name, _ctx_lib) =                \
+	{                                                                      \
+		.mem_slab = &CONCAT(_name, _mem_slab),                         \
+		.mutex = &_name##_mutex                                        \
+	}
+
+/** @brief Context data for a connection. */
+struct bt_conn_ctx {
+	/** Any kind of data associated with a specific connection. */
+	void *data;
+
+	 /** The connection that the data is associated with. */
+	struct bt_conn *conn;
+};
+
+/** @brief Bluetooth connection context library structure. */
+struct bt_conn_ctx_lib {
+	/** Connection contexts. */
+	struct bt_conn_ctx ctx[CONFIG_BT_MAX_CONN];
+
+	/** Context data mutex that ensures that only one connection context is
+	  * manipulated at a time. */
+	struct k_mutex * const mutex;
```

Comment:
I'm not going to block the PR for this, but unless you need priority inversion a `k_sem` does the same with less overhead

Reply:
Thanks @carlescufi. This PR only renamed the library and moved it to a different location in the tree. I had kept separate commits until yesterday, to make it easier to review the actual changes.

@KAGA164 or myself can have a look into using a semaphore instead,
but we'll do that in another PR if that's okay.